x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/TimeSeriesAggregationOperatorFactories.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.compute.operator;
+
+import org.elasticsearch.compute.aggregation.AggregatorFunctionSupplier;
+import org.elasticsearch.compute.aggregation.AggregatorMode;
+import org.elasticsearch.compute.aggregation.GroupingAggregator;
+import org.elasticsearch.compute.aggregation.ValuesBooleanAggregatorFunctionSupplier;
+import org.elasticsearch.compute.aggregation.ValuesBytesRefAggregatorFunctionSupplier;
+import org.elasticsearch.compute.aggregation.ValuesDoubleAggregatorFunctionSupplier;
+import org.elasticsearch.compute.aggregation.ValuesIntAggregatorFunctionSupplier;
+import org.elasticsearch.compute.aggregation.ValuesLongAggregatorFunctionSupplier;
+import org.elasticsearch.compute.aggregation.blockhash.BlockHash;
+import org.elasticsearch.compute.aggregation.blockhash.TimeSeriesBlockHash;
+import org.elasticsearch.compute.data.ElementType;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * This class provides operator factories for time-series aggregations.
+ * A time-series aggregation executes in three stages, deviating from the typical two-stage aggregation.
+ * For example: {@code sum(rate(write_requests)), avg(cpu) BY cluster, time-bucket}
+ *
+ * 1. Initial Stage:
+ * In this stage, a standard hash aggregation is executed, grouped by tsid and time-bucket.
+ * The {@code values} aggregations are added to collect values of the grouping keys excluding the time-bucket,
+ * which are then used for final result grouping.
+ * {@code rate[INITIAL](write_requests), avg[INITIAL](cpu), values[SINGLE](cluster) BY tsid, time-bucket}
+ *
+ * 2. Intermediate Stage:
+ * Equivalent to the final mode of a standard hash aggregation.
+ * This stage merges and reduces the result of the rate aggregations,
+ * but merges (without reducing) the results of non-rate aggregations.
+ * Certain aggregations, such as count_distinct, cannot have their final results combined.
```

Comment:
Why is this the case? Don't we assume that the final grouping fields are included in the tsid? Or, we don't have the mapping between tsid values and the final groups?

Reply:
Is this because the distinct counts per tsid and time bucket aren't a subset of the dimension groups?

Reply:
Ah, I think my wording is confusing. Just to clarify, this isn't a limitation; it's the reason why we still need to keep the intermediate states of non-rate aggregation after the second stage. I will reword this.